### PR TITLE
:bug: (go/v3, go/v4-alpha): fix makefile targets do not re-install binaries when they are found in the project bin directory

### LINKED ISSUE #2843: Makefile: avoid running prerequisites more than once
### What broke? What's expected?

In the Makefile, specifying prerequisites like so:
```Makefile
$(KUSTOMIZE): $(LOCALBIN)
	curl -s $(KUSTOMIZE_INSTALL_SCRIPT) | bash -s -- $(subst v,,$(KUSTOMIZE_VERSION)) $(LOCALBIN)
```
will trigger Kustomize whenever a file is created/changed in `$(LOCALBIN)`

This can easily be fixed with `$(KUSTOMIZE): | $(LOCALBIN)` (note the "|"  in between). More info can be found [here](https://www.gnu.org/software/make/manual/make.html#Prerequisite-Types)

### Reproducing this issue

```bash
kubebuilder init --domain tutorial.kubebuilder.io --repo tutorial.kubebuilder.io/project
make kustomize
make kustomize # it won't run
touch bin/hello
make kustomize # it will run again
```

### KubeBuilder (CLI) Version

3.4.1

### PROJECT version

v3, v2

### Plugin versions

_No response_

### Other versions

_No response_

### Extra Labels

_No response_

### PR DESCRIPTION
**Pull request by FotiadisM**
#### Description

Change the Makefile targets controller-gen and envtest no longer try to download the binary when they can be found in the project/bin directory.

#### Motivation

Closes: #2843

### CODE REVIEW

File: docs/book/src/component-config-tutorial/testdata/project/Makefile [hunk 1]
```diff
@@ -118,15 +118,15 @@ CONTROLLER_TOOLS_VERSION ?= v0.9.0
 KUSTOMIZE_INSTALL_SCRIPT ?= "https://raw.githubusercontent.com/kubernetes-sigs/kustomize/master/hack/install_kustomize.sh"
 .PHONY: kustomize
 kustomize: $(KUSTOMIZE) ## Download kustomize locally if necessary.
-$(KUSTOMIZE): $(LOCALBIN)
+$(KUSTOMIZE): | $(LOCALBIN)
 	test -s $(LOCALBIN)/kustomize || { curl -s $(KUSTOMIZE_INSTALL_SCRIPT) | bash -s -- $(subst v,,$(KUSTOMIZE_VERSION)) $(LOCALBIN); }
```
Comment (camilamacedo86) [inline docs/book/src/component-config-tutorial/testdata/project/Makefile:122]: See that we are testing and checking if kustomize is on the bin?

File: docs/book/src/component-config-tutorial/testdata/project/Makefile [hunk 2]
```diff
@@ -118,15 +118,15 @@ CONTROLLER_TOOLS_VERSION ?= v0.9.0
...
 .PHONY: controller-gen
 controller-gen: $(CONTROLLER_GEN) ## Download controller-gen locally if necessary.
-$(CONTROLLER_GEN): $(LOCALBIN)
+$(CONTROLLER_GEN): | $(LOCALBIN)
 	GOBIN=$(LOCALBIN) go install sigs.k8s.io/controller-tools/cmd/controller-gen@$(CONTROLLER_TOOLS_VERSION)
 
 .PHONY: envtest
 envtest: $(ENVTEST) ## Download envtest-setup locally if necessary.
-$(ENVTEST): $(LOCALBIN)
+$(ENVTEST): | $(LOCALBIN)
```
Comment (camilamacedo86) [inline docs/book/src/component-config-tutorial/testdata/project/Makefile:131]: Could you please fix here using the same approach as above ?

File: docs/book/src/component-config-tutorial/testdata/project/Makefile [hunk 3]
```diff
@@ -118,15 +118,15 @@ CONTROLLER_TOOLS_VERSION ?= v0.9.0
 KUSTOMIZE_INSTALL_SCRIPT ?= "https://raw.githubusercontent.com/kubernetes-sigs/kustomize/master/hack/install_kustomize.sh"
```
Comment (camilamacedo86) [inline docs/book/src/component-config-tutorial/testdata/project/Makefile:118]: About the title could you please rename this one for : 

`:bug: (go/v3): fix makefile targets do not re-install binaries when they are found in the project bin directory` 

Note that we use the title to generate the release notes.

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
Thank you for the contribution 🥇 

Could you just please address the nits and assign the CLA so that we can move forward here?

File: docs/book/src/component-config-tutorial/testdata/project/Makefile (hunk 3 above)
Comment (camilamacedo86) [reply docs/book/src/component-config-tutorial/testdata/project/Makefile:118]: Also, could you please improve the description of the PR with something like:

**Description**

Change the Makefile targets controller-gen and envtest no longer try to download the binary when they can be found in the project/bin directory.  

**Motivation**

Closes: idIssue

**Comment by FotiadisM** [comment]
Just changed my commit to use `test -s $(LOCALBIN)/<binary>` everywhere and removed `|`

**Review by camilamacedo86 — APPROVED** [review]
/ok-to-test

Great work 🥇 
Thank you for the contribution and to raise that.

**Comment by camilamacedo86** [comment]
/hold

@FotiadisM , it is faling because you did not regenerated the testdata dir.
Could you please run `make generate` and squash the commits so that we can get this one merged?

**Review by camilamacedo86 — APPROVED** [review]
/lgtm

File: docs/book/src/component-config-tutorial/testdata/project/Makefile [hunk 4]
```diff
@@ -124,9 +124,9 @@ $(KUSTOMIZE): $(LOCALBIN)
 .PHONY: controller-gen
 controller-gen: $(CONTROLLER_GEN) ## Download controller-gen locally if necessary.
 $(CONTROLLER_GEN): $(LOCALBIN)
-	GOBIN=$(LOCALBIN) go install sigs.k8s.io/controller-tools/cmd/controller-gen@$(CONTROLLER_TOOLS_VERSION)
+	test -s $(LOCALBIN)/controller-gen || GOBIN=$(LOCALBIN) go install sigs.k8s.io/controller-tools/cmd/controller-gen@$(CONTROLLER_TOOLS_VERSION)
```
Comment (ebaron) [inline docs/book/src/component-config-tutorial/testdata/project/Makefile:127]: Hi, I just came across this PR while fixing a similar issue in our operator's Makefile. I was wondering if these test commands should use the variables for each binary, such as: `test -s $(CONTROLLER_GEN)`. If the user overrides the tool variable (e.g. `CONTROLLER_GEN`), then this test will not work.

Comment (FotiadisM) [reply docs/book/src/component-config-tutorial/testdata/project/Makefile:127]: I think you are right, it should probably be changed to
```Makefile
$(CONTROLLER_GEN): $(LOCALBIN)
	test -s $@ || GOBIN=$(LOCALBIN) go install sigs.k8s.io/controller-tools/cmd/controller-gen@$(CONTROLLER_TOOLS_VERSION)

```

**Pull request merged**